pkg/goDB/DBWorkManager.go

```diff
@@ -288,88 +246,93 @@ func (w *DBWorkManager) readBlocksAndEvaluate(workload DBWorkload, resultMap map
 	// Process the workload
 	// The workload consists of timestamps whose blocks we should process.
 	for b, tstamp := range workload.load {
-
-		var (
-			blocks      [ColIdxCount][]byte
-			blockBroken = false
-		)
-
-		for _, colIdx := range query.columnIndizes {
-
-			// Read the block from the file
-			if blocks[colIdx], err = columnFiles[colIdx].ReadBlock(tstamp); err != nil {
-				blockBroken = true
-				w.logger.Warnf("[D %s; B %d] Failed to read %s.gpf: %s", dir, tstamp, columnFileNames[colIdx], err.Error())
-				break
+		select {
+		case <-ctx.Done():
+			w.logger.Infof("[D %s; B %d] Query cancelled. %d/%d blocks processed", dir, tstamp, b, len(workload.load))
+			return nil
+		default:
+			var (
+				blocks      [ColIdxCount][]byte
+				blockBroken bool
+			)
+
+			for _, colIdx := range query.columnIndices {
+
+				// Read the block from the file
+				if blocks[colIdx], err = columnFiles[colIdx].ReadBlock(tstamp); err != nil {
+					blockBroken = true
+					w.logger.Warnf("[D %s; B %d] Failed to read %s.gpf: %s", dir, tstamp, columnFileNames[colIdx], err.Error())
+					break
+				}
 			}
-		}
-
-		if query.hasAttrTime {
-			key.Time = tstamp
-		}
 
-		if query.hasAttrIface {
-			key.Iface = w.iface
-		}
-
-		// Check whether all blocks have matching number of entries
-		numEntries := int(len(blocks[BytesRcvdColIdx]) / 8)
-		for _, colIdx := range query.columnIndizes {
-			l := len(blocks[colIdx])
-			if l/columnSizeofs[colIdx] != numEntries {
-				blockBroken = true
-				w.logger.Warnf("[Bl %d] Incorrect number of entries in file [%s.gpf]. Expected %d, found %d", b, columnFileNames[colIdx], numEntries, l/columnSizeofs[colIdx])
-				break
-			}
-			if l%columnSizeofs[colIdx] != 0 {
-				blockBroken = true
-				w.logger.Warnf("[Bl %d] Entry size does not evenly divide block size in file [%s.gpf]", b, columnFileNames[colIdx])
-				break
+			if query.hasAttrTime {
+				key.Time = tstamp
 			}
-		}
 
-		// In case any error was observed during above sanity checks, skip this whole block
-		if blockBroken {
-			continue
-		}
-
-		// Iterate over block entries
-		for i := 0; i < numEntries; i++ {
-			// Populate key for current entry
-			for _, colIdx := range query.queryAttributeIndizes {
-				copyToKeyFns[colIdx](i, &key, blocks[colIdx])
+			if query.hasAttrIface {
+				key.Iface = w.iface
 			}
 
-			// Check whether conditional is satisfied for current entry
-			var conditionalSatisfied bool
-			if query.Conditional == nil {
-				conditionalSatisfied = true
-			} else {
-				// Populate comparison value for current entry
-				for _, colIdx := range query.conditionalAttributeIndizes {
-					copyToKeyFns[colIdx](i, &comparisonValue, blocks[colIdx])
+			// Check whether all blocks have matching number of entries
+			numEntries := int(len(blocks[BytesRcvdColIdx]) / 8)
```

Comment:
This will create conflicts with #37 (needs careful cross-checking to avoid unnoticed issues)...